### PR TITLE
Containerapp fix ContainerAppClient and ManagedEnvironmentClient async polling

### PR DESCRIPTION
**Pull request by Greedygre**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
Containerapp

File: src/containerapp/azext_containerapp/_clients.py [hunk 1]
```diff
@@ -70,6 +72,53 @@ def poll(cmd, request_url, poll_if_status):  # pylint: disable=inconsistent-retu
...
+
+    animation.tick()
+    r = send_raw_request(cmd.cli_ctx, "GET", request_url)
+
+    while r.status_code in [200] and start < end:
+        time.sleep(POLLING_SECONDS)
+        animation.tick()
+        r = send_raw_request(cmd.cli_ctx, "GET", request_url)
+        response_body = json.loads(r.text)
+        if response_body["status"].lower() in ["failed", "canceled"]:
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/_clients.py:89]: what about there is no status in body?

Comment (Juliehzl) [inline src/containerapp/azext_containerapp/_clients.py:89]: prefer using safe_get()

File: src/containerapp/azext_containerapp/_clients.py [hunk 2]
```diff
@@ -70,6 +72,53 @@ def poll(cmd, request_url, poll_if_status):  # pylint: disable=inconsistent-retu
...
+
+def poll_results(cmd, request_url):  # pylint: disable=inconsistent-return-statements
+    start = time.time()
+    end = time.time() + POLLING_TIMEOUT
+    animation = PollingAnimation()
+
+    animation.tick()
+    r = send_raw_request(cmd.cli_ctx, "GET", request_url)
+
+    while r.status_code in [202] and start < end:
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/_clients.py:111]: what about the status code is error status code like 400?

Comment (Greedygre) [reply src/containerapp/azext_containerapp/_clients.py:111]: if status code not OK , send_raw_request will throw exception.

File: src/containerapp/azext_containerapp/_clients.py [hunk 3]
```diff
@@ -120,14 +171,13 @@ def update(cls, cmd, resource_group_name, name, container_app_envelope, no_wait=
...
-            request_url = url_fmt.format(
-                management_hostname.strip('/'),
-                sub_id,
-                resource_group_name,
-                name,
-                api_version)
-            return poll(cmd, request_url, "inprogress")
+            from azure.cli.core.azclierror import ResourceNotFoundError
+            operation_url = r.headers[HEADER_LOCATION]
+            response = poll_results(cmd, operation_url)
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/_clients.py:176]: what about operation_url is null or empty?

File: src/containerapp/azext_containerapp/_clients.py [hunk 4]
```diff
@@ -453,14 +503,18 @@ def create(cls, cmd, resource_group_name, name, managed_environment_envelope, no
         if no_wait:
             return r.json()
         elif r.status_code == 201:
+            operation_url = r.headers.get(HEADER_AZURE_ASYNC_OPERATION)
+            if operation_url is None:
+                raise AzureResponseError(f"Http response headers lack of necessary property: '{HEADER_AZURE_ASYNC_OPERATION}'")
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/_clients.py:508]: header?

File: src/containerapp/azext_containerapp/_clients.py [hunk 5]
```diff
@@ -89,14 +144,18 @@ def create_or_update(cls, cmd, resource_group_name, name, container_app_envelope
         if no_wait:
             return r.json()
         elif r.status_code == 201:
+            operation_url = r.headers.get(HEADER_AZURE_ASYNC_OPERATION)
+            if not operation_url:
+                raise AzureResponseError(f"Http response headers lack of necessary property: '{HEADER_AZURE_ASYNC_OPERATION}'")
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/_clients.py:149]: Move the check to polling logic

File: src/containerapp/azext_containerapp/_clients.py [hunk 6]
```diff
@@ -120,14 +179,14 @@ def update(cls, cmd, resource_group_name, name, container_app_envelope, no_wait=
...
-            request_url = url_fmt.format(
-                management_hostname.strip('/'),
-                sub_id,
-                resource_group_name,
-                name,
-                api_version)
-            return poll(cmd, request_url, "inprogress")
+            operation_url = r.headers.get(HEADER_LOCATION)
+            if not operation_url:
+                raise AzureResponseError(f"Http response headers lack of necessary property: '{HEADER_LOCATION}'")
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/_clients.py:184]: same as above

**Review by Juliehzl — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**